### PR TITLE
fix(reactions): accept any emoji, drop the shortcode support check

### PR DESCRIPTION
**Pull request by chenjr0719**
## Summary

Reactions rejected any emoji that failed the shortcode regex `^[a-z0-9_+-]{1,32}$` — a raw-unicode emoji (`👍`), uppercase, or a ZWJ sequence — with `invalid reaction shortcode`, even though there is **no** support/registration check behind reactions (an unknown ASCII shortcode is already accepted and written). That blocked legitimate emoji (e.g. a migrated or picker-external one) for no correctness reason.

This drops the character-set "support check" on the reaction path: any emoji is accepted and the client decides renderability.

## Changes

- `pkg/emoji`: add `CanonicalizeReaction` — NFC-normalise + reject empty / over 64 bytes / control chars, with **no** character-set regex. The existing `Canonicalize` (`[a-z0-9_+-]{1,32}` + `IsStandard`) is unchanged and still used by the custom-emoji **upload** path (named ASCII shortcodes), so this is scoped to reactions.
- `history-service` `ReactMessage`: use `CanonicalizeReaction`; the error now fires only on genuinely malformed input (empty / oversized / control), never on an unsupported-but-valid emoji.
- NFC is retained deliberately: it collapses two normalization forms of the same emoji to one storage key, so a client can't create two separate reactions for one emoji.
- `docs/client-api.md` + the request/reply view: the `shortcode` field now documents "any emoji, FE decides renderability; NFC + size only".

## Verification

- Unit: `CanonicalizeReaction` accepts `👍`, flags, ZWJ family sequences, uppercase, ASCII shortcodes; rejects empty / >64 bytes / control chars; NFC collapses equivalent forms. `ReactMessage` validation tests updated (colon/uppercase are no longer rejections; control-char + oversize are). `pkg/emoji` + `history-service` suites green.
- Verified on our dev stack at the wire layer: reacting with a raw `👍` is accepted and the reply echoes `{"shortcode":"👍","action":"added"}`; an ASCII shortcode still works.

## Summary by CodeRabbit

* **New Features**
  * Reactions now support raw Unicode emoji, including multi-character sequences, as well as textual shortcodes.
  * Reaction values are NFC-normalized consistently and can be up to 64 bytes.
  * Textual shortcodes and Unicode emoji remain distinct reaction values.

* **Bug Fixes**
  * Improved validation rejects empty, invisible-only, or oversized values with clearer error messages.

* **Documentation**
  * Updated client API guidance to describe supported formats and validation behavior.

### CODE REVIEW

**Comment by julianshen** [comment]
**1 finding · 1 inline comment posted** — 🟠 1 warning

### Disputed

- `pkg/emoji/emoji.go:69` (`distributed-system`) — The documented client convention cannot guarantee identity across devices or mixed client versions. A legacy client can add `thumbsup`; an upgraded client can later send `👍` intending to toggle the same visible reaction off. Because NFC preserves both as distinct keys, the second request instead adds another entry, splitting counts and potentially showing the same user twice. The same reachable failure occurs with Unicode presentation variants such as `❤` and `❤️`, including replayed historical values. Preserve arbitrary unknown keys, but define one storage identity for known shortcode/Unicode and variation-selector aliases, and during migration look up legacy aliases before deciding whether a toggle adds or removes.

### Local review memory

_Existing unresolved review discussion is retained here and is not reposted as new findings._

- `pkg/emoji/emoji.go:69` — @chenjr0719: By design — reactions are opaque keys with no support/alias table, so a shortcode↔unicode mapping would contradict the accept-any-emoji premise (and couldn't cover non-gemoji shortcodes). Documented in client-api.md + request-reply.md that… ([thread](https://github.com/hmchangw/newchat/pull/325#discussion_r3840746278))
- `pkg/emoji/emoji.go:73` — @chenjr0719: Fixed — added a 256-byte pre-normalization ceiling before norm.NFC.String, so oversized input is rejected before paying for NFC. (dc86d9e9) ([thread](https://github.com/hmchangw/newchat/pull/325#discussion_r3840746284))
- `docs/client-api.md:3854` — @chenjr0719: Fixed — updated client-api.md and request-reply.md: removed the stale invalid-reaction-shortcode/regex error and documented the empty/invisible-value rejection. (dc86d9e9) ([thread](https://github.com/hmchangw/newchat/pull/325#discussion_r3840746291))

<details>
<summary>📒 Files reviewed (7)</summary>

* `docs/client-api.md`
* `docs/client-api/request-reply.md`
* `history-service/internal/service/reactions.go`
* `history-service/internal/service/reactions_test.go`
* `pkg/emoji/emoji.go`
* `pkg/emoji/emoji_test.go`
* `pkg/model/cassandra/reactions.go`

</details>
<details>
<summary>⚙️ Rules run (9)</summary>

* `cassandra`
* `distributed-system`
* `elasticsearch`
* `mongodb`
* `nats`
* `oom`
* `performance-and-concurrency`
* `silent-failure`
* `tgd-review`

</details>

**Review by julianshen — COMMENTED** [review]
tGD inline review

File: pkg/emoji/emoji.go [hunk 1]
```diff
@@ -43,6 +46,42 @@ func Canonicalize(shortcode string) (string, error) {
...
+const ReactionMaxBytes = 64
+
+// CanonicalizeReaction normalizes a reaction emoji for use as a storage/wire key
+// with NO support check: any emoji is accepted and the FE decides renderability.
+// It keeps only two invariants, neither about "is this supported": NFC (so the
+// same emoji in two normalization forms is one storage key, not two reactions)
+// and a byte cap (so a client can't blob-bomb the reactions map). The cap is the
+// only reject; empty is caught upstream. Returns a client-facing BadRequest past
+// the cap so callers can return the error as-is with a friendly message.
+func CanonicalizeReaction(emojiKey string) (string, error) {
```
Comment (julianshen) [inline pkg/emoji/emoji.go:69]: _🏷️ distributed-consistency_ | _🔴 Blocking_ | _`distributed-system`_

**Equivalent reaction representations create different storage identities.**

CanonicalizeReaction preserves every NFC-distinct representation, although the API now presents shortcodes and raw Unicode as alternative ways to specify an emoji. A legacy client can store ReactionKey{Emoji:"thumbsup", UserAccount:u}; another device or upgraded client can then toggle "👍" for the same user and message. Because those values remain different map keys, the second toggle inserts another reaction instead of removing the existing one. The same divergence occurs for common Unicode variants such as "❤" and "❤️". Mixed clients and replayed historical data therefore split counts and can leave one user with duplicate visible reactions. Define a stable storage identity for known shortcode/Unicode and presentation-selector aliases, and perform a compatible lookup or migration of legacy keys before writing the canonical key.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `pkg/emoji/emoji.go` around line 60: CanonicalizeReaction preserves every NFC-distinct representation, although the API now presents shortcodes and raw Unicode as alternative ways to specify an emoji. A legacy client can store ReactionKey{Emoji:"thumbsup", UserAccount:u}; another device or upgraded client can then toggle "👍" for the same user and message. Because those values remain different map keys, the second toggle inserts another reaction instead of removing the existing one. The same divergence occurs for common Unicode variants such as "❤" and "❤️". Mixed clients and replayed historical data therefore split counts and can leave one user with duplicate visible reactions. Define a stable storage identity for known shortcode/Unicode and presentation-selector aliases, and perform a compatible lookup or migration of legacy keys before writing the canonical key.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

File: pkg/emoji/emoji.go [hunk 2]
```diff
@@ -43,6 +46,42 @@ func Canonicalize(shortcode string) (string, error) {
...
+	if !norm.NFC.IsNormalString(emojiKey) {
```
Comment (julianshen) [inline pkg/emoji/emoji.go:73]: _⚡ performance_ | _🟠 Warning_ | _`performance-and-concurrency`_

**The size guard runs only after unbounded normalization work.**

Concurrent clients can submit transport-sized, non-NFC strings. Each request reaches norm.NFC.String before the 64-byte check, forcing O(input size) CPU and an allocation proportional to the entire attacker-controlled value; repeated requests can drive GC pressure and service saturation despite the stated blob guard. Add a small pre-normalization byte ceiling, sized to accommodate valid decomposed forms, while retaining the post-NFC 64-byte check.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `pkg/emoji/emoji.go` around line 61: Concurrent clients can submit transport-sized, non-NFC strings. Each request reaches norm.NFC.String before the 64-byte check, forcing O(input size) CPU and an allocation proportional to the entire attacker-controlled value; repeated requests can drive GC pressure and service saturation despite the stated blob guard. Add a small pre-normalization byte ceiling, sized to accommodate valid decomposed forms, while retaining the post-NFC 64-byte check.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

File: pkg/emoji/emoji.go [hunk 3]
```diff
@@ -43,6 +46,42 @@ func Canonicalize(shortcode string) (string, error) {
...
+	// reaction. Require at least one visible rune.
+	if !hasVisibleRune(emojiKey) {
+		return "", errcode.BadRequest("reaction emoji is required")
+	}
+	return emojiKey, nil
+}
+
+func hasVisibleRune(s string) bool {
+	for _, r := range s {
+		if !unicode.IsSpace(r) && !unicode.IsControl(r) {
```
Comment (julianshen) [inline pkg/emoji/emoji.go]: _🎯 correctness_ | _🟠 Warning_ | _`tgd-review`_

**The visibility guard accepts zero-width-only reaction keys.**

`unicode.IsSpace` and `unicode.IsControl` do not reject format characters such as U+200B ZERO WIDTH SPACE or U+200D ZERO WIDTH JOINER, and combining marks such as U+FE0F also pass. These inputs therefore create the invisible reactions this guard is intended to prevent. Require a graphic, non-space, non-mark base rune and add tests for zero-width and combining-mark-only inputs.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `pkg/emoji/emoji.go` around line 78: `unicode.IsSpace` and `unicode.IsControl` do not reject format characters such as U+200B ZERO WIDTH SPACE or U+200D ZERO WIDTH JOINER, and combining marks such as U+FE0F also pass. These inputs therefore create the invisible reactions this guard is intended to prevent. Require a graphic, non-space, non-mark base rune and add tests for zero-width and combining-mark-only inputs.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

File: docs/client-api.md [hunk 4]
```diff
@@ -3849,7 +3849,7 @@ Toggles a reaction on a message. Any subscribed room member may react — the se
...
+| `shortcode` | string | yes | The reaction emoji — either a bare shortcode without surrounding colons (e.g. `thumbsup`, `acme_party`) or a raw-unicode emoji (e.g. `👍`, a ZWJ sequence, a flag). The server applies **no support check**: it NFC-normalises the value and rejects only a value over 64 bytes (a resource guard, ~20× any real emoji) — it does **not** check the character set, the standard-emoji set, or the site's registered custom emoji. The **FE decides renderability**. Clients typically offer shortcodes from their picker (built-in standard set + the local site's [`emoji.list`](#emojilist--list-a-sites-custom-emoji)), but any emoji is accepted (e.g. a migrated one the picker doesn't list). |
 
 ```json
```
Comment (julianshen) [inline docs/client-api.md:3854]: _📝 documentation_ | _🟠 Warning_ | _`tgd-review`_

**The documented rejection rules omit invisible values.**

The documentation says values over 64 bytes are the only rejection, but `CanonicalizeReaction` also rejects values containing only whitespace or control characters with `reaction emoji is required`. Update this description and the corresponding statement in `docs/client-api/request-reply.md` to document the empty/invisible-value restriction.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `docs/client-api.md` around line 3854: The documentation says values over 64 bytes are the only rejection, but `CanonicalizeReaction` also rejects values containing only whitespace or control characters with `reaction emoji is required`. Update this description and the corresponding statement in `docs/client-api/request-reply.md` to document the empty/invisible-value restriction.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

File: pkg/emoji/emoji.go (hunk 1 above)
Comment (chenjr0719) [reply pkg/emoji/emoji.go:69]: By design — reactions are opaque keys with no support/alias table, so a shortcode↔unicode mapping would contradict the accept-any-emoji premise (and couldn't cover non-gemoji shortcodes). Documented in client-api.md + request-reply.md that clients should send one consistent representation per emoji; keeping the opaque-key behavior. (dc86d9e9)

File: pkg/emoji/emoji.go (hunk 2 above)
Comment (chenjr0719) [reply pkg/emoji/emoji.go:73]: Fixed — added a 256-byte pre-normalization ceiling before norm.NFC.String, so oversized input is rejected before paying for NFC. (dc86d9e9)

File: pkg/emoji/emoji.go (hunk 3 above)
Comment (chenjr0719) [reply pkg/emoji/emoji.go]: Good catch — replaced the whitespace/control guard with a content-rune check (unicode L/N/S). It rejects zero-width (Cf) and combining-mark (M) only keys while accepting real emoji including ZWJ sequences and variation selectors. Added tests for ZWSP, ZWJ, VS16, and a combining-mark-only input. (dc86d9e9)

File: docs/client-api.md (hunk 4 above)
Comment (chenjr0719) [reply docs/client-api.md:3854]: Fixed — updated client-api.md and request-reply.md: removed the stale invalid-reaction-shortcode/regex error and documented the empty/invisible-value rejection. (dc86d9e9)

File: pkg/emoji/emoji.go [hunk 5]
```diff
@@ -43,6 +46,55 @@ func Canonicalize(shortcode string) (string, error) {
...
+	// including ZWJ sequences and variation selectors, always carry a base symbol.
+	if !hasContentRune(emojiKey) {
+		return "", errcode.BadRequest("reaction emoji is required")
+	}
+	return emojiKey, nil
+}
+
+func hasContentRune(s string) bool {
+	for _, r := range s {
+		if unicode.In(r, unicode.L, unicode.N, unicode.S) {
```
Comment (julianshen) [inline pkg/emoji/emoji.go]: _🎯 correctness_ | _🔴 Blocking_ | _`tgd-review`_

**Visible punctuation-only reactions are incorrectly rejected.**

The content check recognizes only letters, numbers, and symbols, so visible punctuation such as `-` and `_` is rejected as though it were invisible. Both were valid one-character shortcodes under the previous `[a-z0-9_+-]{1,32}` contract, and the new documentation promises rejection only when no visible character exists. Include Unicode punctuation in the accepted content categories.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `pkg/emoji/emoji.go` around line 91: The content check recognizes only letters, numbers, and symbols, so visible punctuation such as `-` and `_` is rejected as though it were invisible. Both were valid one-character shortcodes under the previous `[a-z0-9_+-]{1,32}` contract, and the new documentation promises rejection only when no visible character exists. Include Unicode punctuation in the accepted content categories.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

File: docs/client-api.md [hunk 6]
```diff
@@ -3878,7 +3878,7 @@ Toggles a reaction on a message. Any subscribed room member may react — the se
 
 ##### Error response
 
-See [Error envelope](#6-error-envelope-reference). Common errors: `"messageId is required"`, `"shortcode is required"`, `"invalid reaction shortcode"` (malformed: fails `^[a-z0-9_+-]{1,32}$` after NFC), `"message not found"` (also returned when attempting to _add_ a reaction to a soft-deleted message), `"not subscribed to room"`, `"failed to add reaction"`, `"failed to remove reaction"`.
+See [Error envelope](#6-error-envelope-reference). Common errors: `"messageId is required"`, `"shortcode is required"`, `"reaction emoji too large"` (over 64 bytes post-NFC), `"reaction emoji is required"` (no visible character — empty / whitespace / control / zero-width / combining-mark only), `"message not found"` (also returned when attempting to _add_ a reaction to a soft-deleted message), `"not subscribed to room"`, `"failed to add reaction"`, `"failed to remove reaction"`.
```
Comment (julianshen) [inline docs/client-api.md]: _📝 documentation_ | _🔵 Suggestion_ | _`tgd-review`_

**The documented error for an empty shortcode is incorrect.**

`ReactMessage` checks `req.Shortcode == ""` before canonicalization and returns `"shortcode is required"`, as also asserted by `TestHistoryService_ReactMessage_ValidationErrors`. Therefore `"reaction emoji is required"` applies only to non-empty invisible values, not to empty input as this error description states.

<details>
<summary>📝 Committable suggestion</summary>

> ‼️ **IMPORTANT**
> Review this carefully before committing. It is generated from an automated
> review of an untrusted diff — make sure it replaces exactly the highlighted
> lines, is complete, is correctly indented, and actually does what you want.

```suggestion
See [Error envelope](#6-error-envelope-reference). Common errors: `"messageId is required"`, `"shortcode is required"`, `"reaction emoji too large"` (over 64 bytes post-NFC), `"reaction emoji is required"` (a non-empty value with no visible character — whitespace / control / zero-width / combining-mark only), `"message not found"` (also returned when attempting to _add_ a reaction to a soft-deleted message), `"not subscribed to room"`, `"failed to add reaction"`, `"failed to remove reaction"`.
```

</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `docs/client-api.md` around line 3881: `ReactMessage` checks `req.Shortcode == ""` before canonicalization and returns `"shortcode is required"`, as also asserted by `TestHistoryService_ReactMessage_ValidationErrors`. Therefore `"reaction emoji is required"` applies only to non-empty invisible values, not to empty input as this error description states.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

**Review by julianshen — COMMENTED** [review]
tGD inline review

File: docs/client-api/request-reply.md [hunk 7]
```diff
@@ -1280,7 +1280,7 @@ state. Can always **remove** from a soft-deleted message; cannot **add** to one.
 
 #### Errors
 
-`"messageId is required"`, `"shortcode is required"`, `"invalid reaction shortcode"`
+`"messageId is required"`, `"shortcode is required"`, `"reaction emoji too large"`, `"reaction emoji is required"`
```
Comment (julianshen) [inline docs/client-api/request-reply.md]: _📝 documentation_ | _🔵 Suggestion_ | _`tgd-review`_

**The error list retains the obsolete malformed-format description.**

The following line still labels these errors as `(malformed format)`, even though this change removes format validation and replaces it with size and visible-content checks. Remove that stale parenthetical or replace it with descriptions of the two actual validation conditions.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `docs/client-api/request-reply.md` around line 1283: The following line still labels these errors as `(malformed format)`, even though this change removes format validation and replaces it with size and visible-content checks. Remove that stale parenthetical or replace it with descriptions of the two actual validation conditions.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

File: pkg/emoji/emoji.go (hunk 5 above)
Comment (chenjr0719) [reply pkg/emoji/emoji.go]: Good catch — fixed in f4522c35. The guard used letters/numbers/symbols only, which excluded punctuation like _ and - (both valid shortcode chars). Inverted it to the correct definition: visible = not whitespace, not control/format (unicode.C), not mark (unicode.M) — so _/-/punctuation are accepted while zero-width, combining-mark, and whitespace-only keys are still rejected. Added _/-/+1 accepted test cases.

File: docs/client-api.md (hunk 6 above)
Comment (chenjr0719) [reply docs/client-api.md]: Fixed in f4522c35 — clarified that an empty shortcode returns "shortcode is required" (checked in ReactMessage before canonicalization), distinct from "reaction emoji is required" which covers a non-empty but invisible value.

File: docs/client-api/request-reply.md (hunk 7 above)
Comment (chenjr0719) [reply docs/client-api/request-reply.md]: Fixed in f4522c35 — removed the dangling "(malformed format)" from the error list; the reaction path no longer does format validation.

File: docs/client-api.md [hunk 8]
```diff
@@ -3849,7 +3849,7 @@ Toggles a reaction on a message. Any subscribed room member may react — the se
...
+| `shortcode` | string | yes | The reaction emoji — either a bare shortcode without surrounding colons (e.g. `thumbsup`, `acme_party`) or a raw-unicode emoji (e.g. `👍`, a ZWJ sequence, a flag). The server applies **no support check**: it NFC-normalises the value and rejects only (a) a value over 64 bytes (a resource guard, ~20× any real emoji) or (b) a value with **no visible character** — empty, or made only of whitespace / control / zero-width / combining-mark code points. It does **not** check the character set, the standard-emoji set, or the site's registered custom emoji. The **FE decides renderability**. Note: reactions are stored as **opaque keys with no shortcode↔unicode aliasing**, so `thumbsup` and `👍` are two distinct reactions — the FE should send one consistent representation per emoji. Clients typically offer shortcodes from their picker (built-in standard set + the local site's [`emoji.list`](#emojilist--list-a-sites-custom-emoji)), but any emoji is accepted (e.g. a migrated one the picker doesn't list). |
```
[the remark this answers is not in the document]
Comment (chenjr0719) [reply docs/client-api.md]: Fixed in ea8a4ede — updated the ReactRoomEvent.shortcode and ReactionDelta.shortcode field docs to the canonical NFC-normalised reaction key (textual shortcode or raw-unicode emoji; opaque, not format-validated). The media-service custom-emoji shortcode fields keep their format description since that path still validates.

File: pkg/emoji/emoji.go [hunk 9]
```diff
@@ -43,6 +46,56 @@ func Canonicalize(shortcode string) (string, error) {
...
+	// real emoji (ZWJ sequences / variation selectors carry a visible base char).
+	if !hasVisibleRune(emojiKey) {
+		return "", errcode.BadRequest("reaction emoji is required")
+	}
+	return emojiKey, nil
+}
+
+func hasVisibleRune(s string) bool {
+	for _, r := range s {
+		if !unicode.IsSpace(r) && !unicode.In(r, unicode.C, unicode.M) {
```
Comment (julianshen) [inline pkg/emoji/emoji.go]: _🎯 correctness_ | _🟠 Warning_ | _`tgd-review`_

**Aggregate Unicode C filtering rejects renderable characters.**

`unicode.C` includes private-use (`Co`) and currently unassigned (`Cn`) code points, not only controls and format characters. This rejects private-use glyphs and emoji introduced after the backend's Unicode tables even when the frontend can render them, contradicting the stated frontend-decides policy. Restrict the invisible check to control (`Cc`) and format (`Cf`) categories, and add a private-use regression test.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
In `pkg/emoji/emoji.go` around line 92: `unicode.C` includes private-use (`Co`) and currently unassigned (`Cn`) code points, not only controls and format characters. This rejects private-use glyphs and emoji introduced after the backend's Unicode tables even when the frontend can render them, contradicting the stated frontend-decides policy. Restrict the invisible check to control (`Cc`) and format (`Cf`) categories, and add a private-use regression test.

Fix only this issue, keep the change minimal, and make sure the tests still pass.
```

</details>

**Review by julianshen — COMMENTED** [review]
tGD inline review

File: pkg/emoji/emoji.go (hunk 9 above)
Comment (chenjr0719) [reply pkg/emoji/emoji.go]: Good catch — fixed in 08ca9222. Narrowed the denylist from the whole C category to Cc+Cf (control + format) plus M (marks): unicode.C also covers private-use (Co) and unassigned (Cn), so a private-use glyph or a future emoji not yet in this build's Unicode tables was being wrongly rejected. Now only truly-invisible control/format/mark (and whitespace) are rejected; private-use and not-yet-assigned code points are accepted, consistent with FE-decides-renderability. Added a private-use accepted test.

File: docs/client-api.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (chenjr0719) [reply docs/client-api.md]: Fixed in 47a89f0e — reworded the field doc to state the full domain: any non-empty, visible textual or Unicode key is accepted (bare shortcode, colon-wrapped, ASCII punctuation, raw-unicode emoji, or private-use glyph), with the examples now illustrative rather than exhaustive.

**Review by mliu33 — APPROVED** [review]
